chain/client/src/sync/epoch.rs

```diff
@@ -0,0 +1,508 @@
+
+        let mut update = chain.mut_chain_store().store_update();
+        update.save_block_header_no_update_tree(last_header.clone())?;
+        update.save_block_header_no_update_tree(
+            proof.current_epoch.last_block_header_in_prev_epoch,
+        )?;
+        update.save_block_header_no_update_tree(
+            proof.current_epoch.second_last_block_header_in_prev_epoch,
+        )?;
+        tracing::info!(
+            "last final block of last past epoch: {:?}",
+            proof.past_epochs.last().unwrap().last_final_block_header.hash()
+        );
+        update.save_block_header_no_update_tree(
+            proof.past_epochs.last().unwrap().last_final_block_header.clone(),
+        )?;
+        update.force_save_header_head(&Tip::from_header(&last_header))?;
+        update.save_final_head(&Tip::from_header(&self.genesis))?;
+
+        epoch_manager.init_after_epoch_sync(
+            &mut store_update,
+            proof.last_epoch.first_block_in_epoch,
+            proof.last_epoch.second_last_block_in_epoch,
+            proof.last_epoch.last_block_in_epoch.clone(),
+            proof.last_epoch.last_block_in_epoch.epoch_id(),
+            proof.last_epoch.epoch_info,
+            proof.last_epoch.final_block_header_in_next_epoch.epoch_id(),
+            proof.last_epoch.next_epoch_info,
+            proof.last_epoch.final_block_header_in_next_epoch.next_epoch_id(),
+            proof.last_epoch.next_next_epoch_info,
+        )?;
+
+        store_update.insert_ser(
+            DBCol::BlockInfo,
+            &borsh::to_vec(proof.current_epoch.first_block_info_in_epoch.hash()).unwrap(),
+            &proof.current_epoch.first_block_info_in_epoch,
+        )?;
+
+        store_update.set_ser(
+            DBCol::BlockOrdinal,
+            &borsh::to_vec(&proof.current_epoch.merkle_proof_for_first_block.size()).unwrap(),
+            last_header.hash(),
+        )?;
+
+        store_update.set_ser(
+            DBCol::BlockHeight,
+            &borsh::to_vec(&last_header.height()).unwrap(),
+            last_header.hash(),
+        )?;
+
+        store_update.set_ser(
+            DBCol::BlockMerkleTree,
+            last_header.hash().as_bytes(),
+            &proof.current_epoch.merkle_proof_for_first_block,
+        )?;
+
+        update.merge(store_update);
+        update.commit()?;
+
+        *status = SyncStatus::EpochSyncDone;
+
+        Ok(())
+    }
+}
+
+impl Handler<EpochSyncRequestMessage> for ClientActorInner {
+    #[perf]
+    fn handle(&mut self, msg: EpochSyncRequestMessage) {
+        let store = self.client.chain.chain_store.store().clone();
+        let network_adapter = self.client.network_adapter.clone();
+        let route_back = msg.route_back;
+        self.client.epoch_sync.async_computation_spawner.spawn(
+            "respond to epoch sync request",
+            move || {
+                let proof = match EpochSync::derive_epoch_sync_proof(store) {
```

Comment:
I don't want to do that, because in apply_proof, the mutable store is passed in. If we stored the Store in the class, it creates creates ambiguity in which store to use.